core/src/main/java/org/elasticsearch/common/util/concurrent/AbstractLifecycleRunnable.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.common.util.concurrent;
+
+import org.elasticsearch.common.component.Lifecycle;
+import org.elasticsearch.common.logging.ESLogger;
+
+import java.util.Objects;
+
+/**
+ * {@code AbstractLifecycleRunnable} is a service-lifecycle aware {@link AbstractRunnable}.
+ * <p>
+ * This simplifies the running and rescheduling of {@link Lifecycle}-based {@code Runnable}s.
+ */
+public abstract class AbstractLifecycleRunnable extends AbstractRunnable {
+    /**
+     * The monitored lifecycle for the associated service.
+     */
+    private final Lifecycle lifecycle;
+    /**
+     * The service's logger (note: this is passed in!).
+     */
+    private final ESLogger logger;
+
+    /**
+     * {@link AbstractLifecycleRunnable} must be aware of the actual {@code lifecycle} to react properly.
+     *
+     * @param lifecycle The lifecycle to react too
+     * @param logger The logger to use when logging
+     * @throws NullPointerException if any parameter is {@code null}
+     */
+    public AbstractLifecycleRunnable(Lifecycle lifecycle, ESLogger logger) {
+        Objects.requireNonNull(lifecycle, "lifecycle must not be null");
+        Objects.requireNonNull(logger, "logger must not be null");
+
+        this.lifecycle = lifecycle;
```

Comment:
`Objects.requireNonNull()` either returns the first argument or throws an exception so you could simplify this to:

```java
this.lifecycle = Objects.requireNonNull(lifecycle, "lifecycle must not be null");
this.logger = Objects.requireNonNull(logger, "logger must not be null");
```